File: solutions/scala/secret-handshake/2/src/main/scala/SecretHandshake.scala
object SecretHandshake:
  private val signs =
    Map(1 -> "wink", 2 -> "double blink", 4 -> "close your eyes", 8 -> "jump")

  def commands(i: Int): List[String] =
    val s = signs.collect { case (k, v) if (i & k) != 0 => v }.toList
    if (i & 16) != 0 then s.reverse else s

File: solutions/haskell/secret-handshake/2/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits (Bits ((.&.)), shiftL)
import Data.List (foldl')

actionMap :: [(Int, [String] -> [String])]
actionMap =
  [ (shiftL 1 0, \acc -> acc ++ ["wink"]),
    (shiftL 1 1, \acc -> acc ++ ["double blink"]),
    (shiftL 1 2, \acc -> acc ++ ["close your eyes"]),
    (shiftL 1 3, \acc -> acc ++ ["jump"]),
    (shiftL 1 4, reverse)
  ]

handshake :: Int -> [String]
handshake n = foldl' (\acc (mask, apply) -> if n .&. mask /= 0 then apply acc else acc) [] actionMap